File: src/rust/day05.rs
#[derive(Debug)]
pub struct Range {
    src_start: usize,
    src_end: usize,
    dst_start: usize,
}

impl Range {
    pub fn new(src_start: usize, src_end: usize, dst_start: usize) -> Range {
        Range {
            src_start,
            src_end,
            dst_start,
        }
    }

    pub fn get(&self, i: usize) -> Option<usize> {
        if i >= self.src_start && i < self.src_end {
            let relative_offset = i - self.src_start;
            Some(self.dst_start + relative_offset)
        } else {
            None
        }
    }
}

#[derive(Debug)]
pub struct Mapping {
    name: String,
    ranges: Vec<Range>,
}

impl Mapping {
    pub fn new(name: String) -> Mapping {
        Mapping {
            name,
            ranges: Vec::new(),
        }
    }

    pub fn update(&mut self, s: &str) {
        let parts: Vec<usize> = s
            .split_whitespace()
            .map(|x| x.parse::<usize>().unwrap())
            .collect();
        let (dst, src, rng) = (parts[0], parts[1], parts[2]);
        let r = Range::new(src, src + rng, dst);
        self.ranges.push(r);
    }

    pub fn get(&self, i: usize) -> usize {
        for r in &self.ranges {
            if let Some(new) = r.get(i) {
                return new;
            }
        }
        i
    }
}

#[derive(Debug)]
pub struct Pipeline {
    maps: Vec<Mapping>,
}

impl Pipeline {
    pub fn new() -> Pipeline {
        Pipeline { maps: Vec::new() }
    }

    pub fn from_lines(lines: Vec<String>) -> Pipeline {
        let mut pipeline = Pipeline::new();
        let mut current_map: Option<Mapping> = None;

        for line in lines {
            let trimmed_line = line.trim();
            if trimmed_line.ends_with("map:") {
                if let Some(map) = current_map {
                    pipeline.maps.push(map);
                }
                let name = trimmed_line.split_whitespace().next().unwrap().to_string();
                current_map = Some(Mapping::new(name));
            } else if trimmed_line.is_empty() {
                if let Some(map) = current_map.take() {
                    pipeline.maps.push(map);
                }
            } else {
                if let Some(ref mut map) = current_map {
                    map.update(trimmed_line);
                } else {
                    panic!(
                        "Got a parseable line but no Mapping obj to update: {}",
                        trimmed_line
                    );
                }
            }
        }

        if let Some(map) = current_map {
            pipeline.maps.push(map);
        }

        pipeline
    }

    pub fn get(&self, i: usize, debug: bool) -> usize {
        let mut new_i = i;
        for map in &self.maps {
            new_i = map.get(new_i);
            if debug {
                println!("{} -> {} -> {}", i, map.name, new_i);
            }
        }
        new_i
    }
}

File: python/aoc_2023/day05.py
from dataclasses import dataclass, field
from typing import List


@dataclass
class Range:
    src_start: int
    src_end: int
    offset: int  # can be negative

    def get(self, i: int):
        if i >= self.src_start and i <= self.src_end:
            return i + self.offset


@dataclass
class Mapping:
    name: str
    ranges: List[Range] = field(default_factory=list)

    def update(self, s: str):
        "take an individual line like 0 15 37 and update the internal map"
        dst, src, rng = [int(i) for i in s.split()]
        r = Range(src_start=src, src_end=src + rng, offset=dst - src)
        self.ranges.append(r)

    def get(self, i: int):
        for r in self.ranges:
            if new := r.get(i):
                return new
        return i


@dataclass
class Pipeline:
    maps: List[Mapping]

    @classmethod
    def from_lines(cls, lines: List[str]):
        "Parse a text blob to create a series of mappings"
        maps = []
        m = None
        for line in lines:
            line = line.strip()
            if line.endswith("map:"):
                name = line.split()[0]
                m = Mapping(name=name)
            elif not line:
                maps.append(m)
                m = None
            else:
                if m is None:
                    raise RuntimeError(
                        "got a parseable line but no Mapping obj to update: %s", line
                    )
                m.update(line)
        maps.append(m)
        return cls(maps=maps)

    def get(self, i: int, debug=False):
        "Return the final value as i goes through each mapping in the pipeline"
        for mapping in self.maps:
            new = mapping.get(i)
            if debug:
                print(f"{i} -> {mapping.name} -> {new}")
            i = new
        return i
